### ISSUE TITLE
TypeScript allows class properties to reference instance variables that have not been defined yet.

### ISSUE THREAD

**Issue by wgoodall01** [microsoft/TypeScript#31223]
**TypeScript Version:**  3.4.5 (and whatever runs in the playground)

**Search Terms:**
 - class property
 - property initialization

**Code**

```ts
class A{
    num: number;
    constructor(num:number) {
        this.num = num;
    }

    computed = this.num * 10;
}

const instance = new A();
console.log(instance.computed); // gives NaN
```

**Expected behavior:**

Some kind of type error when we try to assign `computed`. Although `computed` is part of the type signature of `A`, we know that the property has not been assigned yet, because class properties are evaluated before the body of the constructor.

**Actual behavior:**

No such error. If you run the example, it prints `NaN`, as `this.num` is `undefined` when `computed` is defined (before the body of the constructor has run). 

The example compiles to the following JS:

```js
var A = /** @class */ (function () {
    function A(num) {
        this.computed = this.num * 10;
        this.num = num;
    }
    return A;
}());
var instance = new A(3);
console.log(instance.computed); // gives NaN
```

[**Playground Link:**](https://www.typescriptlang.org/play/#src=class%20A%7B%0D%0A%20%20%20%20num%3A%20number%3B%0D%0A%20%20%20%20constructor(num%3Anumber)%20%7B%0D%0A%20%20%20%20%20%20%20%20this.num%20%3D%20num%3B%0D%0A%20%20%20%20%7D%0D%0A%0D%0A%20%20%20%20computed%20%3D%20this.num%20*%2010%3B%0D%0A%7D%0D%0A%0D%0Aconst%20instance%20%3D%20new%20A(3)%3B%0D%0Aconsole.log(instance.computed)%3B%20%2F%2F%20gives%20NaN) 

**Related Issues:**

 - #14488 and #8476 (quite similar). However, in the spirit of the strict-mode check implemented in #20075, it might be good to check the same in this case.

**Comment by wgoodall01** [comment]
Interesting addition: if, instead of using an assignment inside the constructor, you use a parameter property to set the instance variable, it works just fine.

For instance, this TypeScript:

```ts
class A{
    constructor(public num:number) {}
    computed = this.num * 10;
}

const instance = new A(3);
console.log(instance.computed); // gives 30
```

will compile to this:

```js
var A = /** @class */ (function () {
    function A(num) {
        this.computed = this.num * 10;
        this.num = num;
    }
    return A;
}());
var instance = new A(3);
console.log(instance.computed); // gives NaN
```
which works fine. ([playground link](https://www.typescriptlang.org/play/#src=class%20A%7B%0D%0A%20%20%20%20num%3A%20number%3B%0D%0A%20%20%20%20constructor(num%3Anumber)%20%7B%0D%0A%20%20%20%20%20%20%20%20this.num%20%3D%20num%3B%0D%0A%20%20%20%20%7D%0D%0A%0D%0A%20%20%20%20computed%20%3D%20this.num%20*%2010%3B%0D%0A%7D%0D%0A%0D%0Aconst%20instance%20%3D%20new%20A(3)%3B%0D%0Aconsole.log(instance.computed)%3B%20%2F%2F%20gives%20NaN))

Again, I'm not quite sure if this is a compiler bug, or rather just a check that was never implemented.

**Comment by j-oliveras** [comment]
Duplicate of #31158?

**Comment by wgoodall01** [comment]
It's the same idea, although this case the order of evaluation of the different parts of the constructor (parameter properties, then class properties, then constructor body) allows you to write code that isn't type safe.

**Comment by weswigham** [comment]
![image](https://user-images.githubusercontent.com/2932786/57409522-5a484280-719d-11e9-8a0a-ec3839401959.png)

We should _probably_ issue an error on using uninitialized instance fields in other field initializers, and also error on referencing following (and therefore uninitialized) fields. Although you can technically safely write `class A { a = this.c = 0; b = this.c; c = 12;  }`, however odd that would be.

**Comment by DanielRosenwasser** [comment]
Why is this slated for 3.6.3?

**Comment by rbuckton** [comment]
It was still sitting in the 3.6.1 milestone and was a fairly quick fix.

**Comment by evmar** [comment]
@RyanCavanaugh  I think this was fixed in TS 4.0.

**Comment by rbuckton** [comment]
This was fixed by #38030.
